### ISSUE TITLE
How could I use several breadcrumbs on page?

### ISSUE THREAD

**Issue by bocharsky-bw** [whiteoctober/BreadcrumbsBundle#46]
There are any solution to use a few different breadcumbs on the same page - main breadcrumbs and for sub block one?

**Comment by richsage** [comment]
Great idea! There's nothing that supports that presently - maybe some kind of namespace would work? eg:

    $breadcrumbs->getNamespace("main")->addBreadcrumb(...);

or

    $breadcrumbs->addNamespaceItem("main", $txt, $url);

with a default namespace of some kind. I'm open to suggestions :-)

**Comment by bocharsky-bw** [comment]
@richsage Thanks for quick reply!

I like first case, because with second case it's look like we need to add 2 methods: `addNamespaceItem` and `addNamespaceRouteItem` and it will be a bit awkwardly as for me.

I also considered to implement a `factory` pattern for this or it will be too complex? What do you think?

**Comment by richsage** [comment]
OK, I can do the first one. So, we can assume that `$breadcrumbs->addBreadcrumb()` uses the default namespace, and then if namespaces are required then the separate `getNamespace` method needs to be called? Maybe the standalone `addBreadcrumb` method then needs to be deprecated so as to force namespace usage in the future?

What did you have in mind for the factory pattern? I suspect as you say that it might be a bit complex, but interested to see your approach.

**Comment by bocharsky-bw** [comment]
I considered [KnpMenuBundle](http://symfony.com/doc/master/bundles/KnpMenuBundle/index.html) solution to build a few menus from factory service. In [Menu as a service](http://symfony.com/doc/master/bundles/KnpMenuBundle/menu_service.html) described how to use  factory method to build menus. For example, in order to breadcumbs:
``` yaml
services:
    app.breadcrumbs.builder:
        class: App\Breadcrumbs\Builder
        arguments: ["@wo.breadcrumbs.builder"]
        # where "@wo.breadcrumbs.builder" is a core breadcrumbs service from WhiteOctoberBreadcrumbsBundle

    app.breadcrumbs.main:
        class: App\Breadcrumbs\MainBreadcrumbs
        factory_service: wo.breadcrumbs.builder
        factory_method: createMainBreadcrumbs
        arguments: []
        tags:
            - { name: wo_breadcrumbs, alias: main_breadcrumbs }

    app.breadcrumbs.sub_breadcrumb:
        class: App\Breadcrumbs\SubBreadcrumbs
        factory_service: wo.breadcrumbs.builder
        factory_method: createSubBreadcrumbs
        arguments: []
        tags:
            - { name: wo_breadcrumbs, alias: sub_breadcrumbs }
```

Also we can add default crumbs to it, for example, when I create `main_breadcrumbs` breadcrumbs it already should contain home crumb (we avoid to add home crumb at every action).

I thinks this could entail BC, but we can avoid it to add a new twig function to render breadcrumbs from factory.

@richsage What do you think?

**Comment by richsage** [comment]
Right I see - with you now. I think that would be a useful addition; the separate twig function would be good to avoid a BC break. I'm keen that the existing methods can still be used for more "lightweight" usage.

What's your thoughts on the namespace approach above as well then - in keeping the `addBreadcrumb` method but assuming a default namespace? Is that too confusing?

**Comment by bocharsky-bw** [comment]
No, it isn't. I think namespace approach with default namespace is more simply as factory in my opinion, and we will provide this new feature.

**Comment by richsage** [comment]
@bocharsky-bw I've started work on this at the branch [namespace-crumbs](https://github.com/whiteoctober/BreadcrumbsBundle/tree/namespace-crumbs).

**Comment by bocharsky-bw** [comment]
@richsage It's great! I'll try to test it

**Comment by bocharsky-bw** [comment]
The `wo_is_final_breadcrumb` filter looks broken. The `WhiteOctober\BreadcrumbsBundle\Model\Breadcrumbs::count` method counting namespaces, not crumbs now.

**Comment by richsage** [comment]
I'll pick that up now...

**Comment by richsage** [comment]
@bocharsky-bw OK that should be fixed now

**Comment by bocharsky-bw** [comment]
I think it will be great to throw a nice exception message when try to use nonexistent namespace.
Now I get something like:
```
Key "nonexistent_namespace" in object with ArrayAccess of class "WhiteOctober\BreadcrumbsBundle\Model\Breadcrumbs" does not exist in WhiteOctoberBreadcrumbsBundle::breadcrumbs.html.twig at line 4
```
What do you think?

**Comment by richsage** [comment]
Yes that would be nice! I'll adjust now

**Comment by richsage** [comment]
We also need to add some documentation :-)

**Comment by bocharsky-bw** [comment]
Ok, I'll add some docs about namespace

**Comment by bocharsky-bw** [comment]
ping @richsage

**Comment by bocharsky-bw** [comment]
I think we would to have a possible **BC** break with `Iterator`, `ArrayAccess` and `Countable` interfaces implementation in `WhiteOctober\BreadcrumbsBundle\Model\Breadcrumbs` if any used this implemented methods directly in their code or overridden `breadcrumbs.html.twig` template with their own. The behavior of this methods was changed.
Is it a problem?

**Comment by richsage** [comment]
@bocharsky-bw ah I see what you mean - iterating just gives the namespace keys, not the actual crumbs... hmm, tricky one. Let me ponder that...

**Comment by bocharsky-bw** [comment]
@richsage Yes, it is. OK!

**Comment by bocharsky-bw** [comment]
@richsage I think I fixed this BC break in #57 (kept default behavior and added new functional in order to use breadcrumbs namespaces). What do you think?

**Comment by richsage** [comment]
I need to test that out I think - will come back to you :-)

**Comment by bocharsky-bw** [comment]
Yeah, it's good idea, thanks! :-)

**Comment by richsage** [comment]
OK so #57 is in now, I had a play and it seems good. Are we good to go with this now?

**Comment by bocharsky-bw** [comment]
Yes, I am. I don't have more any idea what else we could do there.

**Comment by richsage** [comment]
PR opened - #62 :-)

**Comment by bocharsky-bw** [comment]
Thanks!

**Comment by richsage** [comment]
Released in [1.2.0](https://github.com/whiteoctober/BreadcrumbsBundle/releases/tag/1.2.0)

**Comment by bocharsky-bw** [comment]
Awesome, thank you!
